tools/chpl-language-server/src/chpl-language-server.py

```diff
@@ -0,0 +1,331 @@
+
+    assert isinstance(node, chapel.core.NamedDecl)
+    return node.name()
+
+
+def decl_kind(decl: chapel.core.NamedDecl) -> Optional[SymbolKind]:
+    if isinstance(decl, chapel.core.Module) and decl.kind() != "implicit":
+        return SymbolKind.Module
+    elif isinstance(decl, chapel.core.Class):
+        return SymbolKind.Class
+    elif isinstance(decl, chapel.core.Record):
+        return SymbolKind.Struct
+    elif isinstance(decl, chapel.core.Interface):
+        return SymbolKind.Interface
+    elif isinstance(decl, chapel.core.Enum):
+        return SymbolKind.Enum
+    elif isinstance(decl, chapel.core.EnumElement):
+        return SymbolKind.EnumMember
+    elif isinstance(decl, chapel.core.Function):
+        if decl.is_method():
+            return SymbolKind.Method
+        elif decl.name() in ("init", "init="):
+            return SymbolKind.Constructor
+        elif decl.kind() == "operator":
+            return SymbolKind.Operator
+        else:
+            return SymbolKind.Function
+    elif isinstance(decl, chapel.core.Variable) and decl.is_field():
+        return SymbolKind.Field
+    elif isinstance(decl, chapel.core.Variable):
+        if decl.is_field():
+            return SymbolKind.Field
+        elif decl.intent() == "<const-var>":
+            return SymbolKind.Constant
+        elif decl.intent() == "type":
+            return SymbolKind.TypeParameter
+        else:
+            return SymbolKind.Variable
+    return None
+
+
+def get_symbol_information(
+    decl: chapel.core.NamedDecl, uri: str
+) -> Optional[SymbolInformation]:
+    loc = Location(uri, location_to_range(decl.location()))
+    if kind := decl_kind(decl):
+        # TODO: should we use DocumentSymbol or SymbolInformation
+        # LSP spec says prefer DocumentSymbol, but nesting doesn't work out of the box. implies that we need some kind of visitor pattern to build a DS tree
+        # using symbol information for now, as it sort-of autogets the tree structure
+        is_deprecated = chapel.is_deprecated(decl)
+        return SymbolInformation(loc, decl.name(), kind, deprecated=is_deprecated)
+    return None
+
+
+@dataclass
+class NodeAndRange:
+    node: chapel.core.AstNode
+    rng: Range = field(init=False)
+
+    def __post_init__(self):
+        self.rng = location_to_range(self.node.location())
+
+    def get_location(self):
+        path = self.node.location().path()
+        return Location(f"file://{path}", self.rng)
+
+
+@dataclass
+class ResolvedPair:
+    ident: NodeAndRange
+    resolved_to: NodeAndRange
+
+
+@dataclass
+class FileInfo:
+    uri: str
+    context: chapel.core.Context
+    segments: List[ResolvedPair] = field(default_factory=list)
+    siblings: chapel.SiblingMap = field(init=False)
+
+    def __post_init__(self):
+        self.rebuild_segments()
+
+    def parse_file(self) -> List[chapel.core.AstNode]:
+        """
+        Given a file URI, return the ASTs making up that file. Advances
```

Comment:
The file URI is not given -- it's part of the class. Also, docstrings on other methods perhaps?